### DISCUSSION TITLE
[Question]: How to make it work with Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by Harrenkyym** [danny-avila/LibreChat#1567] [Q&A]
### What is your question?

Hi. I'm trying to make LibreChat work in my NAS. 

That's what I would expect to do to make it work in NPM, as I do with any other service running in my NAS.

![Sin título2](https://github.com/danny-avila/LibreChat/assets/76946315/41bb8885-5583-47e4-aef1-44825537372e)

But I don't know how to set my .env:

HOST=0.0.0.0 (?)
PORT=3080

DOMAIN_CLIENT= (?)
DOMAIN_SERVER= (?)

Thanks!

### What is the main subject of your question?

Installation

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Did you check out the nginx.conf file in the client folder?

**Reply by Harrenkyym** [reply]
Hmmmmm...

I don't know exactly what to check there (I'm not the best one checking for this things, I'm just a tryharder trying to figure out things).
How that affects my config? Should I change something? Could you guide me? I use Nginx Proxy Manager for every service and it works like a charm. You know, for duckdns, etc. I would like to use it aswell for LibreChat.

**Reply by Harrenkyym** [reply]
The thing is that I have 2 DDNS (one free and another paid with a cooler domain name, but I want to keep my free one, because the paid one sometimes have problems and the free one is smooth).

So I've checked this: https://github.com/danny-avila/LibreChat/discussions/890
and I don't really know how to make it work. Because I'd like to make work my internal IP 192.168.X.X (static one) so with NPM I can do the reverse proxy for both of my domains, like in the image from my original question.

Ports 80 and 443 are already being used by other services (like NPM itself) and ports are open. I don't really know how to make the approach.

And if I check the logs:

>LibreChat@0.6.5 backend
 cross-env NODE_ENV=production node api/server/index.js
/app/node_modules/file-stream-rotator/FileStreamRotator.js:672
                        throw e;
                        ^
Error: EACCES: permission denied, mkdir '/app/api/logs/'
    at Object.mkdirSync (node:fs:1391:3)
    at /app/node_modules/file-stream-rotator/FileStreamRotator.js:669:24
    at Array.reduce (<anonymous>)
    at mkDirForFile (/app/node_modules/file-stream-rotator/FileStreamRotator.js:656:27)
    at FileStreamRotator.getStream (/app/node_modules/file-stream-rotator/FileStreamRotator.js:532:5)
    at new DailyRotateFile (/app/node_modules/winston-daily-rotate-file/daily-rotate-file.js:80:57)
    at Object.<anonymous> (/app/api/config/winston.js:43:3)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32) {
  errno: -13,
  syscall: 'mkdir',
  code: 'EACCES',
  path: '/app/api/logs/'
}
Node.js v18.19.0

**Reply by danny-avila** [reply]
> Error: EACCES: permission denied, mkdir '/app/api/logs/'

Maybe there's a file permission issue that's causing the server to crash on startup?

**Reply by Harrenkyym** [reply]
And how do I solve it? I already understood there was a problem with file permissions.

Anyway, how do I config the .env file and the nginx.conf file?

**Reply by danny-avila** [reply]
You may want to see if this guide is relevant for you: https://docs.librechat.ai/deployment/digitalocean.html#part-v-editing-the-nginx-file-for-custom-domains-and-advanced-configs

TL;DR use the `deploy-compose.yml` file instead of default `docker-compose.yml`, which uses a prebuilt image from github containers. this has NGINX configured, and you can follow the guide I linked to edit nginx.

**Reply by Harrenkyym** [reply]
Yesterday I edited the nginx.conf like this:

server_name librechat.*;

I guess I won't access trough: [my nas internal ip]:3080 (that is something that I would like, because I have this setup for every other docker container I'm running), but I still not know what do I have to do with (because I'm using 2 ddns, I have tried one, just to test, and it doesn't work anyways):

HOST=0.0.0.0 (?)
PORT=3080

DOMAIN_CLIENT= (?)
DOMAIN_SERVER= (?)

**Reply by danny-avila** [reply]
Note that I've never owned a NAS or setup this project on a NAS but here's what I think you should set these values as

HOST=empty or your NAS IP
PORT=3080

DOMAIN_CLIENT= the domain, for example including http or https. note that you should run it in HTTPS and this can be done via nginx conf or cloudflare tunneling. I can't help you with this.
DOMAIN_SERVER= same as client

here's my real example as I use this setup
```bash
# .env

# HOST= # I leave this field empty as the default value "localhost" works for me
PORT=3080

DOMAIN_CLIENT=https://custom.domain.ai
DOMAIN_SERVER=https://custom.domain.ai
```

my nginx (I have cloudflare tunneling for HTTPS)

```bash
# client/nginx.conf

server {
  listen 80;

  server_name custom.domain.ai;

   # Increase the client_max_body_size to allow larger file uploads
   # The default limits for image uploads as of 11/22/23 is 20MB/file, and 25MB/request
  client_max_body_size 25M;

  location /api {
    proxy_pass http://api:3080/api;
  }

  location / {
    proxy_pass http://api:3080;
  }
}

```

 I spin up the containers with the following command:
 `docker-compose -f ./deploy-compose.yml up`

**Reply by danny-avila** [reply]
Cloudflare guide: https://docs.librechat.ai/deployment/cloudflare.html#installation-steps

**Reply by Harrenkyym** [reply]
I already have 2 https domains: https://librechat.custom.duckdns.org and https://librechat.custom.com

Both are being redirected to my NAS internal IP:port. But obviously my .env and nginx.conf aren't well configurated. I have a question, would it be possible to have something like this?

HOST=0.0.0.0
PORT=3080

DOMAIN_CLIENT=https://librechat.custom.duckdns.org,https://librechat.custom.com
DOMAIN_SERVER=https://librechat.custom.duckdns.org,https://librechat.custom.com

and in my .env file: server_name librechat.*; so it works for both? I've seen this wildcard here https://github.com/danny-avila/LibreChat/discussions/890

**Reply by danny-avila** [reply]
> I have a question, would it be possible to have something like this?

Probably, but this is beyond my expertise. I say try to get something working and then experiment or consult AI on your setup.

**Reply by Harrenkyym** [reply]
Hey, Danny. I told you I'm a tryharder. I'm a fucking tryharder.

I did it. It's running.

If anyone ever asks again:

My problem was the folder api/logs/ wasn't being created. So after many tries (tried to explicit it in the ```Dockerfile```/```Dockerfile.multi```, tried to gave permissions, tried to set UID and GID for the api in the ```deploy-compose```, etc.) I decided to create the folder manually then map the volume in the ```deploy-compose```: ```- ./api/logs:/app/api/logs```

Then, my ```.env``` (I modified many things apart from that)

```
DOMAIN_CLIENT="https://librechat.custom.duckdns.org https//librechat.custom.com"
DOMAIN_SERVER="https://librechat.custom.duckdns.org https//librechat.custom.com"
```

And my ```nginx.conf```:


```
server {

  listen 80;
  listen 443 ssl;

  # ssl_certificate /etc/nginx/ssl/nginx.crt;
  # ssl_certificate_key /etc/nginx/ssl/nginx.key;

  server_name librechat.*;

   # Increase the client_max_body_size to allow larger file uploads
   # The default limits for image uploads as of 11/22/23 is 20MB/file, and 25MB/request
  client_max_body_size 25M;

  location /api {
    proxy_pass http://api:3080/api;
  }

  location / {
    proxy_pass http://api:3080;
  }
}
```

I think I haven't forgot anything.

Now is when I tell you that ```my internal ip:port``` is working. So I can proxy reverse it with NPM. And now is when I tell you too, that I have tried both ddns at the same time and they are working. Perfect! But I thought... what if I try a third one that I have not specified in the ```.env``` file and the subdomain is not even ```librechat``` as I set in my ```nginx.conf```? Something like https//gpt.custom.com... It works too! So I'm happy, but I'm pretty sure that part of the code of Librechat is useless. Because if I have to set both domain client and server to supposedly make it work, and it works even if I'm using a different one... I don't even know if the syntax I'm using for two domains is correct. I asked ChatGPT and that was the answer, but I can't confirm it.

Well, it's working. :)

**Reply by danny-avila** [reply]
Amazing!! So sorry you had to deal with that pesky folder permission issue. I also forgot to mention that DOMAIN_CLIENT/SERVER is only relevant for OAuth logins, forgive me as it's been 7 months since I've worked with those values hands on. 

Happy you got it setup and hope it serves you well, man.

**Reply by CXwudi** [reply]
So does it means that if I disabled the OAuth, I can comment out DOMAIN_CLIENT/SERVER?

**Comment by dltj** [comment]
> My problem was the folder api/logs/ wasn't being created. So after many tries (tried to explicit it in the `Dockerfile`/`Dockerfile.multi`, tried to gave permissions, tried to set UID and GID for the api in the `deploy-compose`, etc.) I decided to create the folder manually then map the volume in the `deploy-compose`: `- ./api/logs:/app/api/logs`

@Harrenkyym : This is where I get stuck...how did you create the folder manually? Did you need to add `RUN mkdir /app/api/logs` and `RUN /app/client/public/images` to a Dockerfile and rebuild?

**Reply by Harrenkyym** [reply]
Man, I use Windows on my laptop, and I have access to some shared folders of my NAS, so I literally created it manually, I didn't use any command.

**Reply by dltj** [reply]
Ah, thanks. I ended up creating the directories in the clone, and those got copied into the Docker image. That'll work for my purposes.
